File: sim.py
from math import *
from visualize import Visualization
import time
import threading


class Simulation(threading.Thread):
    """docstring for Simulation"""

    def __init__(self):
        threading.Thread.__init__(self)
        '''
        Ask/define the parameters
        '''
        self.yf1 = 0
        self.pyf1 = 0
        self.fixj = 0
        self.pfixj = 0
        self.ci = 0
        self.tetr = 0
        self.a = 0
        self.b = 0
        self.c = 0
        self.promoter1 = 1
        self.promoter2 = 0
        self.promoterA = 1
        self.promoterB = 0
        self.promoterC = 0
        self.rbs1 = 1
        self.rbs2 = 0.75
        self.dePhosCoeff1 = 0.5
        self.dePhosCoeff2 = 0.5
        self.phosp = 1
        self.blue_intensity = 0
        self.red_intensity = 0  # for the intensity switch
        self.degCoeffYF1 = 1
        self.degCoeffFixJ = 1
        self.degCoeffCI = 1
        self.degCoeffTetR = 1
        self.degCoeffA = 0.5
        self.degCoeffB = 0.5
        self.degCoeffC = 0.5
        self.timeStep = 0.1
        self.iterations = 600
        self.data = {
            'YF1': [self.yf1],
            'PYF1': [self.pyf1],
            'FixJ': [self.fixj],
            'PFixJ': [self.pfixj],
            'CI': [self.ci],
            'TetR': [self.tetr],
            'A': [self.a],
            'B': [self.b],
            'C': [self.c]
        }
        self.timesteps = [0]
        self.visualization = Visualization(self)
        self.visualization.start()

    def getAmount(self, proteinName):
        '''
        Get the current amount of selected protein
        '''
        return self.data.get(proteinName)[len(self.data.get(proteinName)) - 1]

    def derivativeYF1(self, currentConcentration):
        return self.promoter1 * self.rbs1 + \
            self.dePhosCoeff1 * self.getAmount('PYF1') - \
            (self.degCoeffYF1 + self.blue_intensity) * currentConcentration

    def derivativePYF1(self, currentConcentration):
        return self.blue_intensity * self.getAmount('YF1') - \
            (self.dePhosCoeff1 + self.degCoeffYF1) * currentConcentration

    def derivativeFixJ(self, currentConcentration):
        return self.promoter1 * self.rbs1 + \
            self.dePhosCoeff2 * self.getAmount('PFixJ') - \
            (self.phosp * self.getAmount('PYF1') + self.degCoeffFixJ) * currentConcentration

    def derivativePFixJ(self, currentConcentration):
        return self.phosp * self.getAmount('FixJ') * self.getAmount('PYF1') - \
            (self.dePhosCoeff2 + self.degCoeffFixJ) * currentConcentration

    def derivativeCI(self, currentConcentration):
        return self.promoter2 * self.rbs1 - self.degCoeffCI * currentConcentration

    def derivativeTetR(self, currentConcentration):
        return (self.promoterA + self.promoterB) * self.rbs2 - \
            self.degCoeffTetR * currentConcentration

    def derivativeA(self, currentConcentration):
        return self.promoterA * self.rbs1 - self.degCoeffA * currentConcentration

    def derivativeB(self, currentConcentration):
        return self.promoterB * self.rbs1 - self.degCoeffB * currentConcentration

    def derivativeC(self, currentConcentration):
        return self.promoterC * self.rbs1 - self.degCoeffC * currentConcentration

    def derivativeSelect(self, proteinName, currentConcentration):
        '''
        Select appropriate derivative
        '''
        if proteinName == 'YF1':
            return self.derivativeYF1(currentConcentration)
        elif proteinName == 'PYF1':
            return self.derivativePYF1(currentConcentration)
        elif proteinName == 'FixJ':
            return self.derivativeFixJ(currentConcentration)
        elif proteinName == 'PFixJ':
            return self.derivativePFixJ(currentConcentration)
        elif proteinName == 'CI':
            return self.derivativeCI(currentConcentration)
        elif proteinName == 'TetR':
            return self.derivativeTetR(currentConcentration)
        elif proteinName == 'A':
            return self.derivativeA(currentConcentration)
        elif proteinName == 'B':
            return self.derivativeB(currentConcentration)
        elif proteinName == 'C':
            return self.derivativeC(currentConcentration)

    def promoterUpdate(self):
        self.promoter2 = 7 * self.getAmount('PFixJ')
        self.promoterA = (1 - self.red_intensity) * (1 - self.getAmount('CI'))
        if self.getAmount('CI') < 1:
            self.promoterB = (1 - self.red_intensity) * (self.getAmount('CI'))
        else:
            self.promoterB = (1 - self.red_intensity) * (1 - 2.5 * (self.getAmount('CI') - 1))
        self.promoterC = (1 - self.red_intensity) * (1 - (1.75 / 0.75) * self.getAmount('TetR'))

    def run(self):
        '''
        Runge-Kutta computation for protein concentrations
        '''
        for i in range(self.iterations):
            for key in self.data:
                x = self.getAmount(key)
                coeff1 = self.derivativeSelect(key, x)
                coeff2 = self.derivativeSelect(
                    key, (x + (coeff1) * self.timeStep / 2))
                coeff3 = self.derivativeSelect(
                    key, (x + (coeff2) * self.timeStep / 2))
                coeff4 = self.derivativeSelect(
                    key, (x + (coeff3) * self.timeStep))
                x = x + (self.timeStep / 6) * (coeff1 + 2 * coeff2 + 2 * coeff3 + coeff4)
                if(x > 0):
                    self.data.get(key).append(x)
                else:
                    self.data.get(key).append(0)
            self.timesteps.append(i + 1)
            self.promoterUpdate()
            self.visualization.update()
        time.sleep(60)

sim = Simulation()

File: WebUI/sim.js
    var yf1 = 0;
    var pyf1 = 0;
    var fixj = 0;
    var pfixj = 0;
    var ci = 0;
    var tetr = 0;
    var a = 0;
    var b = 0;
    var c = 0;
    var promoter1 = 1;
    var promoter2 = 0;
    var promoterA = 1;
    var promoterB = 0;
    var promoterC = 0;
    var rbs1 = 1;
    var rbs2 = 0.75;
    var dePhosCoeff1 = 0.5;
    var dePhosCoeff2 = 0.5;
    var phosp = 1;
    var blue_intensity = 0;
    var red_intensity = 0;  // for the intensity switch
    var degCoeffYF1 = 1;
    var degCoeffFixJ = 1;
    var degCoeffCI = 1;
    var degCoeffTetR = 1;
    var degCoeffA = 0.5;
    var degCoeffB = 0.5;
    var degCoeffC = 0.5;
    var timeStep = 0.1;
    var iterations = 600;
    var data = {"YF1":[yf1],
                "PYF1":[pyf1],
                "FixJ":[fixj],
                "PFixJ":[pfixj],
                "CI":[ci],
                "TetR":[tetr],
                "A":[a],
                "B":[b],
                "C":[c]
            }
    var timesteps = [0]; // ??
    //var visualization = Visualization(self) // ??
    //var visualization.start() // ??

function getAmount(proteinName){
    return data[proteinName][data[proteinName].length -1];
}

function derivativeYF1(currentConcentration){
    return promoter1 * rbs1 + dePhosCoeff1 * getAmount('PYF1') - (degCoeffYF1 + blue_intensity) * currentConcentration;
}

function derivativePYF1(currentConcentration){
    return blue_intensity * getAmount('YF1') - (dePhosCoeff1 + degCoeffYF1) * currentConcentration;
}

function derivativeFixJ(currentConcentration){
    return promoter1 * rbs1 + dePhosCoeff2 * getAmount('PFixJ') - (phosp * getAmount('PYF1') + degCoeffFixJ) * currentConcentration;
}

function derivativePFixJ(currentConcentration){
    return phosp * getAmount('FixJ') * getAmount('PYF1') - (dePhosCoeff2 + degCoeffFixJ) * currentConcentration;
}

function derivativeCI(currentConcentration){
    return promoter2 * rbs1 - degCoeffCI * currentConcentration;
}

function derivativeTetR(currentConcentration){
    return (promoterA + promoterB) * rbs2 - degCoeffTetR * currentConcentration;
}

function derivativeA(currentConcentration){
    return promoterA * rbs1 - degCoeffA * currentConcentration;
}

function derivativeB(currentConcentration){
    return promoterB * rbs1 - degCoeffB * currentConcentration;
}

function derivativeC(currentConcentration){
    return promoterC * rbs1 - degCoeffC * currentConcentration;
}

function derivativeSelect(proteinName, currentConcentration){
    //Select appropriate derivative
    if (proteinName == 'YF1'){
        return derivativeYF1(currentConcentration);
    }
    else if (proteinName == 'PYF1'){
        return derivativePYF1(currentConcentration);
    }
    else if (proteinName == 'FixJ'){
        return derivativeFixJ(currentConcentration);
    }
    else if (proteinName == 'PFixJ'){
        return derivativePFixJ(currentConcentration);
    }
    else if (proteinName == 'CI'){
        return derivativeCI(currentConcentration);
    }
    else if (proteinName == 'TetR'){
        return derivativeTetR(currentConcentration);
    }
    else if (proteinName == 'A'){
        return derivativeA(currentConcentration);
    }
    else if (proteinName == 'B'){
        return derivativeB(currentConcentration);
    }
    else if (proteinName == 'C'){
        return derivativeC(currentConcentration);
    }
}

function promoterUpdate(){
    promoter2 = 7 * getAmount('PFixJ');
    promoterA = (1 - red_intensity) * (1 - getAmount('CI'));
    if (getAmount('CI') < 1){
        promoterB = (1 - red_intensity) * (getAmount('CI'));
    }
    else{
        promoterB = (1 - red_intensity) * (1 - 2.5 * (getAmount('CI') - 1));
    }
    promoterC = (1 - red_intensity) * (1 - (1.75 / 0.75) * getAmount('TetR'));
}

function oneStep(i){
    var keyList = Object.keys(data);

    for(var j = 0; j < keyList.length;j++){
        x = getAmount(keyList[j]);
        coeff1 = derivativeSelect(keyList[j], x);
        coeff2 = derivativeSelect(
            keyList[j], (x + (coeff1) * timeStep / 2));
        coeff3 = derivativeSelect(
            keyList[j], (x + (coeff2) * timeStep / 2));
        coeff4 = derivativeSelect(
            keyList[j], (x + (coeff3) * timeStep));
        x = x + (timeStep / 6) * (coeff1 + 2 * coeff2 + 2 * coeff3 + coeff4);
        if(x > 0){
            data[(keyList[j])].push(x);
        }
        else{
            data[(keyList[j])].push(0);
        }
    }
    timesteps.push(i + 1);
    promoterUpdate();
    var R = 200 + (getAmount('A') * 27.5) - (getAmount('B') * 45) - (getAmount('C') * 45);
    var G = 185 + (getAmount('B') * 35) - (getAmount('A') * 40) - (getAmount('C') * 40);
    var B = 125 + (getAmount('C') * 65) - (getAmount('B') * 25) - (getAmount('C') * 25);
    // A = red, B = green, C = blue. They reduce the others to make the colors more vivid.
    // The base values for each color are for a nice E.Coli light brown.
    setColonyColor("RGB(" + R  + ","  + G + "," + B + ")");
}

function run(){
    //Runge-Kutta computation for protein concentrations
    var i = 0;
    setInterval(function() {
        var light = getLightIntensities();
        blue_intensity = light.blue/100;
        red_intensity = 1 - light.red/100;
        //console.log(blue_intensity);
        //console.log(red_intensity);
        i++;
        oneStep(i);
    }, 100)
}

run();
